### PR TITLE
Add command to revoke tokens (https://github.com/ansible/awx/issues/2030)

### LINKED ISSUE #2030: Make awx-command for Revoking All Tokens
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

 - API

##### SUMMARY
1. Make an awx-command for revoking all OAuth2 tokens.
2. Add an optional `--user` flag to revoke a single user's OAuth2 tokens
3. Make an optional `--revoke_refresh` flag to also revoke the associated refresh token

##### ADDITIONAL INFORMATION

```
# The following command should revoke all existing OAuth2 Tokens 
$ awx-manage revoke_tokens

# The following command should revoke all OAuth2 Tokens & their refresh tokens
$ awx-manage revoke_tokens --revoke_refresh

# The following command should revoke all OAuth2 Tokens for the user with id=<pk>. 
$ awx-manage revoke_tokens --user=<pk>

# The following command should revoke all OAuth2 Tokens & refresh token for the user with id=<pk>. 
$ awx-manage revoke_tokens --user=<pk> --revoke_refresh

# <pk-dne> is an user that does not exist.
$ awx-manage revoke_tokens --user=<pk-dne>
CommandError: The user does not exist.
```

### PR DESCRIPTION
**Pull request by cdvv7788**
Signed-off-by: Cristian Vargas <cristian@swapps.co>

##### SUMMARY

Fixes https://github.com/ansible/awx/issues/2030
I added a command as required. It allows to revoke access/refresh tokens globally and per user.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

 - API

##### AWX VERSION

```
awx: 1.0.6.24
```

##### ADDITIONAL INFORMATION

The logic with the Refresh tokens can use some review (check that the model revoked is the right one).
According to https://github.com/ansible/awx/issues/2030 there should be 2 commands, one named revoke_tokens and another named expire_sessions. While that does make sense for keeping intent explicit, adding a different command for doing something that the same command should be able to handle may be harder to recall and brings not much benefit (someone using the commands should already know that tokens are tied to sessions, and that revoking them should expire said sessions).
I would like to discuss that detail.

```

```

### CODE REVIEW

**Comment by rooftopcellist** [comment]
@cdvv7788 I made an error in the issue, all of those example commands should have said `revoke_tokens`.  I have fixed the issue.  I am pulling down your PR and looking at it.

**Comment by rooftopcellist** [comment]
It looks like you've got some flake8 tests failing you'll have to fix, but that should be quick:

```./awx/main/management/commands/revoke_tokens.py:10:1: E302 expected 2 blank lines, found 1
./awx/main/management/commands/revoke_tokens.py:14:1: E302 expected 2 blank lines, found 1
```

File: awx/main/management/commands/revoke_tokens.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+                revoke_tokens(RefreshToken.objects.all())
+            revoke_tokens(OAuth2AccessToken.objects.all())
+        else:
+            try:
+                user = User.objects.get(username=options['user'])
+            except ObjectDoesNotExist:
+                raise CommandError('The user does not exist.')
+            if options['revoke_refresh']:
+                revoke_tokens(RefreshToken.objects.filter(user=user))
+            revoke_tokens(user.main_oauth2accesstoken.filter(user=user))
```
Comment (rooftopcellist) [inline awx/main/management/commands/revoke_tokens.py]: It looks like when you call `token.revoke()` on a RefreshToken, django-oauth-toolkit (DOT) attempts to delete the related AccessToken ([here](https://github.com/jazzband/django-oauth-toolkit/blob/master/oauth2_provider/models.py#L380).  If the related AccessToken doesn't exist (which is possible as of DOT 1.1.2), we will run into the following strace.  

```
[root@awx /]# awx-manage revoke_tokens --user gandalf --revoke_refresh
Traceback (most recent call last):
  File "/usr/local/bin/awx-manage", line 9, in <module>
    load_entry_point('awx', 'console_scripts', 'awx-manage')()
  File "/awx_devel/awx/__init__.py", line 116, in manage
    execute_from_command_line(sys.argv)
  File "/venv/awx/lib/python2.7/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/venv/awx/lib/python2.7/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/venv/awx/lib/python2.7/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/venv/awx/lib/python2.7/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/awx_devel/awx/main/management/commands/revoke_tokens.py", line 33, in handle
    revoke_tokens(RefreshToken.objects.filter(user=user))
  File "/awx_devel/awx/main/management/commands/revoke_tokens.py", line 12, in revoke_tokens
    token.revoke()
  File "/venv/awx/lib/python2.7/site-packages/oauth2_provider/models.py", line 371, in revoke
    access_token_model.objects.get(id=self.access_token_id).revoke()
  File "/venv/awx/lib/python2.7/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/venv/awx/lib/python2.7/site-packages/django/db/models/query.py", line 380, in get
    self.model._meta.object_name
awx.main.models.oauth.DoesNotExist: OAuth2AccessToken matching query does not exist.
```

One proposed workaround until this is fixed in DOT, is to amend your `revoke_token()` function to just delete the AWX token resources.

**Review by rooftopcellist — CHANGES_REQUESTED** [review]
@cdvv7788 Great work here!  I'd like to see some tests for this before merging.

**Comment by cdvv7788** [comment]
@rooftopcellist  I modified the file a little. Is that what you meant?

I tried running the tests but it was raising an error: pluggy.PluginValidationError: Plugin 'timeout' could not be loaded

I will rebase and try to add the tests again once we confirm that this works.

**Comment by rooftopcellist** [comment]
@cdvv7788 It looks like the issue with not being able to revoke refresh tokens in DOT is now fixed with this commit: https://github.com/ansible/awx/commit/87d6940158910c3e24c447f435851b344911d3fb

We forked DOT to fix it and then pinned our fork in AWX for now until it is fixed upstream.  As a result, we shouldn't need the custom `revoke()` method you wrong now.  

-----------------------------------

I tested this out, and found an issue: when you revoke all tokens with --revoke_refresh, it only revokes the tokens, not their refresh_tokens.  Are you seeing that as well?

```
awx-manage revoke_tokens --revoke_refresh
```

Also, re: tests, that looks like a pluggy test failure due to an old test dependency.  I'll look into that some more.

File: awx/main/management/commands/revoke_tokens.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+from django.core.exceptions import ObjectDoesNotExist
+from django.utils import timezone
+
+
+# AWX
+from awx.main.models.oauth import OAuth2AccessToken
+from oauth2_provider.models import RefreshToken
+
+
+def revoke(token):
```
Comment (rooftopcellist) [inline awx/main/management/commands/revoke_tokens.py]: We should be able to do without this revoke() method now that [this issue](https://github.com/jazzband/django-oauth-toolkit/issues/625) is fixed in our fork of DOT.

**Review by rooftopcellist — CHANGES_REQUESTED** [review]

**Comment by rooftopcellist** [comment]
@cdvv7788 I fixed that test failure, the next push you make to this branch will kick of integration tests again and it should pass.

**Comment by ryanpetrello** [comment]
@rooftopcellist what's the state of this one?

**Comment by cdvv7788** [comment]
@ryanpetrello I need to revert the last change and add some tests. I will try to make some time this weekend to finish what is missing.

**Comment by rooftopcellist** [comment]
recheck

**Comment by rooftopcellist** [comment]
This looks great!  I have tested it out locally with PAT's and application tokens and revoking access_tokens and refresh_tokens works as I would expect:

```
bash-4.2$ awx-manage revoke_oauth2_tokens
revoked OAuth2AccessToken 6VBPcQSwbKYd9iih3LCBSlUjCApr2S
revoked OAuth2AccessToken jMmpqffsQfZKItZ068ptyNGmce6rs1
bash-4.2$ awx-manage revoke_oauth2_tokens --all
revoked RefreshToken BTwAePraxkiKWNmDzTGewluXs78Ms9
revoked RefreshToken mGSKPx4UJHWFkEBMs0epHsE9sptxXb
```

The refresh_token hangs around, but is set to revoked, which is how the latest DOT handles this.  

```
awx-dev=> select * from main_oauth2accesstoken;
 id | token | expires | scope | created | updated | description | last_used | application_id | user_id | source_refresh_token_id | modified
----+-------+---------+-------+---------+---------+-------------+-----------+----------------+---------+-------------------------+----------
(0 rows)

awx-dev=> select * from oauth2_provider_refreshtoken;
 id |             token              | access_token_id | application_id | user_id |            created            |            updated            |            revoked
----+--------------------------------+-----------------+----------------+---------+-------------------------------+-------------------------------+-------------------------------
  4 | BTwAePraxkiKWNmDzTGewluXs78Ms9 |                 |              1 |       1 | 2018-11-08 15:00:20.9484+00   | 2018-11-08 15:00:55.591004+00 | 2018-11-08 15:00:55.59064+00
  5 | mGSKPx4UJHWFkEBMs0epHsE9sptxXb |                 |              1 |       1 | 2018-11-08 15:00:29.187923+00 | 2018-11-08 15:00:55.603439+00 | 2018-11-08 15:00:55.603118+00
(2 rows)
```

**Review by rooftopcellist — APPROVED** [review]
This LGTM.  Lets shipit!

**Comment by cdvv7788** [comment]
Nice! Sorry about the lack of updates. Thanks for finishing it @ryanpetrello

**Comment by ryanpetrello** [comment]
@cdvv7788 thanks for the PR - I just helped push it across the finish line :)!

**Pull request merged**